src/cursor/cur_version.c

```diff
@@ -0,0 +1,470 @@
+        } else {
+            page = cbt->ref->page;
+
+            /* Retrieve the value for each type of underlying table structure. */
+            switch (page->type) {
+            case WT_PAGE_ROW_LEAF:
+                rip = &page->pg_row[cbt->slot];
+                __wt_row_leaf_value_cell(session, cbt->ref->page, rip, vpack);
+                break;
+            case WT_PAGE_COL_FIX:
+            case WT_PAGE_COL_VAR:
+                recno = cbt->ref->ref_recno;
+                cip = &page->pg_var[cbt->slot];
+                cell = WT_COL_PTR(page, cip);
+                __wt_cell_unpack_kv(session, page->dsk, cell, vpack);
+
+                if (vpack->type == WT_CELL_DEL) {
+                    F_SET(version_cursor, WT_VERSION_CUR_ON_DISK_EXHAUSTED);
+                    ret = WT_NOTFOUND;
+                    goto done;
+                }
+                break;
+            default:
+                WT_ERR(__wt_illegal_value(session, page->type));
+                break;
+            }
+
+            if (page->type == WT_PAGE_COL_FIX)
+                WT_IGNORE_RET(
+                  __wt_col_fix_get_time_window(session, cbt->ref, cbt->ref->ref_recno, &vpack->tw));
+
+            if (!WT_TIME_WINDOW_HAS_STOP(&vpack->tw)) {
+                durable_stop_ts = version_cursor->upd_durable_stop_ts;
+                stop_ts = version_cursor->upd_stop_ts;
+                stop_txn = version_cursor->upd_stop_txnid;
+            } else {
+                durable_stop_ts = vpack->tw.durable_stop_ts;
+                stop_ts = vpack->tw.stop_ts;
+                stop_txn = vpack->tw.stop_txn;
+            }
+
+            __wt_cursor_set_key(cursor, vpack->tw.start_txn, vpack->tw.start_ts,
+              vpack->tw.durable_start_ts, stop_txn, stop_ts, durable_stop_ts, WT_UPDATE_STANDARD,
+              tw.prepare, 0, WT_VERSION_DISK_IMAGE);
+            cursor->value.data = vpack->data;
+            cursor->value.size = vpack->size;
+            F_SET(cursor, WT_CURSTD_VALUE_INT);
+
+            upd_found = true;
+            F_SET(version_cursor, WT_VERSION_CUR_ON_DISK_EXHAUSTED);
+        }
+    }
+
+    if (!upd_found && !F_ISSET(version_cursor, WT_VERSION_CUR_HS_EXAUSTED)) {
+        /*
+         * If the history store cursor is not yet positioned, then we are traversing the history
+         * store versions for the first time.
+         */
+        if (!F_ISSET(hs_cursor, WT_CURSTD_KEY_INT)) {
+            hs_cursor->set_key(hs_cursor, 4, S2BT(session)->id, cursor->key, WT_TS_MAX, UINT64_MAX);
+            WT_ERR_NOTFOUND_OK(__wt_curhs_search_near_before(session, hs_cursor), true);
+        } else
+            WT_ERR_NOTFOUND_OK(ret = hs_cursor->prev(hs_cursor), true);
+
+        /*
+         * If there are no history store records for the given key or if we have iterated through
+         * all the records already, we have exhausted the history store.
+         */
+        if (ret == 0) {
```

Comment:
No need to check return.

Reply:
I thought we still needed to check because ret may not be 0 if the key is not matching, but not sure if that is classified as an error or not.

Reply:
No worries, you are right, ret should be 0 by here.

Reply:
If key is not match, hs search returns not found. Changing WT_ERR_NOTFOUND_OK to WT_ERR handles this case.